pyppeteer/page.py

```diff
@@ -943,30 +914,21 @@ def set_request(req: Request) -> None:
             Usage of the History API to change the URL is considered a
             navigation.
         """  # noqa: E501
-        options = merge_dict(options, kwargs)
-        mainFrame = self._frameManager.mainFrame
-        if mainFrame is None:
-            raise PageError('No main frame.')
-        timeout = options.get('timeout', self._defaultNavigationTimeout)
-        watcher = NavigatorWatcher(self._frameManager, mainFrame, timeout,
-                                   options)
-        responses: Dict[str, Response] = dict()
-        listener = helper.addEventListener(
-            self._networkManager,
-            NetworkManager.Events.Response,
-            lambda response: responses.__setitem__(response.url, response)
-        )
-        result = await watcher.navigationResponse()
-        helper.removeEventListeners([listener])
-        error = result[0].pop().exception()
-        if error:
-            raise error
-
-        response = responses.get(self.url, None)
-        return response
-
-    async def waitForRequest(self, urlOrPredicate: Union[str, Callable[[Request], bool]],  # noqa: E501
-                             options: Dict = None, **kwargs: Any) -> Request:
+        return await self.mainFrame.waitForNavigation(
+            timeout=timeout, waitUntil=waitUntil)
+
+    def _sessionClosePromise(self):
+        # TODO implement this
+        if not self._disconnectPromise:
+            self._disconnectPromise = self._loop.create_future()
+            self._client.once(Events.CDPSession.Disconnected)
```

Comment:
This is my first guess:
```suggestion
            self._client.once(Events.CDPSession.Disconnected, lambda: self._disconnectPromise.set_exception(PageError('Target Closed'))
```